src/ray/common/asio/instrumented_io_context.h

```diff
@@ -61,6 +61,23 @@ struct GuardedGlobalStats {
   mutable absl::Mutex mutex;
 };
 
+/// An opaque stats handle, used to manually instrument event loop handlers that don't
+/// hook into a .post() call.
+struct StatsHandle {
+  std::string handler_name;
+  int64_t start_time;
+  std::shared_ptr<GuardedHandlerStats> handler_stats;
+  std::shared_ptr<GuardedGlobalStats> global_stats;
+
+  StatsHandle(std::string handler_name_, int64_t start_time_,
+              std::shared_ptr<GuardedHandlerStats> handler_stats_,
+              std::shared_ptr<GuardedGlobalStats> global_stats_)
+      : handler_name(handler_name_),
+        start_time(start_time_),
+        handler_stats(handler_stats_),
+        global_stats(global_stats_) {}
+};
```

Comment:
Could we RAY_CHECK(false) on destruction if the handle isn't processed by RecordExecution prior to destruction? This would avoid leaking stats by accident.

Reply:
Would that entail adding some kind of `execution_recorded` flag that `RecordExecution` sets and checking that in the destructor?

Reply:
Yep!

Reply:
Hmm I just implemented this, and it appears that this won't behave well with the deadline timers. If the periodical runner is destructed or the deadline timer is cancelled, the callback will never be invoked and the opaque stats handle will be destructed without `RecordExecution` having been called.

Reply:
I implemented it in [this commit](https://github.com/ray-project/ray/pull/15144/commits/62cf8fca670d78c1160f0a9526b6cbe6e3a25725) if you want to take a look, I'm going to revert the commit for now.

Reply:
`deadline timer is cancelled` isn't it something we should handle? How will the stats look like if this happens ?

Reply:
Definitely, I'm working on preventing deadline timer leaks in case the timer is cancelled or destructed before triggering.

Reply:
@rkooo567 @ericl I took the approach of manually cleaning up the queueing stats if handler execution was never recorded. If/as the queueing stats expand, we can isolate that queue stats logic in a static function on `instrumented_io_context` that will be shared by `RecordExecution()` and the `StatsHandle` destructor. Let me know if y'all think of a cleaner approach!

https://github.com/ray-project/ray/pull/15144/commits/63a60465125c604bbf7eb998966971f039604637